File: scripts/algorithms/M/Minimum Area Rectangle II/Minimum Area Rectangle II.js
// Runtime: 98 ms (Top 66.67%) | Memory: 61.40 MB (Top 66.67%)

var minAreaFreeRect = function(points) {
    const diagonalMap = new Map();
    let result = Infinity;

    for (let a = 0; a < points.length - 1; a++) {
        const [x1, y1] = points[a];
        for (let b = a + 1; b < points.length; b++) {
            const [x2, y2] = points[b];
            const diagonal = calculateeDiagonal(x1, y1, x2, y2);
            const centerX = (x1 + x2) / 2;
            const centerY = (y1 + y2) / 2;
            const key = `${diagonal}_${centerX}_${centerY}`;
            const vectors = diagonalMap.get(key) ?? [];

            vectors.push({ x1, x2, y1, y2 });
            diagonalMap.set(key, vectors);
        }
    }

    for (const [key, vector] of diagonalMap) {
        if (vector.length < 2) continue;
        for (let a = 0; a < vector.length - 1; a++) {
            const vectorA = vector[a];
            for (let b = a + 1; b < vector.length; b++) {
                const vectorB = vector[b];
                const w = Math.sqrt(calculateeDiagonal(vectorA.x1, vectorA.y1, vectorB.x1, vectorB.y1));
                const h = Math.sqrt(calculateeDiagonal(vectorA.x1, vectorA.y1, vectorB.x2, vectorB.y2));

                result = Math.min(result, w * h);
            }
        }
    }
    return result === Infinity ? 0 : result;

    function calculateeDiagonal(x1, y1, x2, y2) {
        return Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2);
    }
};

File: scripts/algorithms/M/Minimum Area Rectangle II/Minimum Area Rectangle II.java
// Runtime: 64 ms (Top 66.67%) | Memory: 56.7 MB (Top 16.67%)
class Solution {
    public double minAreaFreeRect(int[][] points) {
        Map<Pair<Double, Double>, Map<Double, List<int[][]>>> map = new HashMap<>();

        double res = Double.MAX_VALUE;
        for (int i = 0; i < points.length; i++) {
            int[] p1 = points[i];
            for (int j = i+1; j < points.length; j++) {
                int[] p2 = points[j];
                // get mid point
                Pair<Double, Double> pm = new Pair((p1[0]+p2[0])/2d, (p1[1]+p2[1])/2d);
                if (!map.containsKey(pm))
                    map.put(pm, new HashMap<>());
                // get diagonal length
                double dist2 = dist2(p1, p2);
                if (!map.get(pm).containsKey(dist2))
                    map.get(pm).put(dist2, new ArrayList<>());

                // calculate area for each pair of p3/p4 and check min
                // Worst case is each pair has same mid point with same length
                // At worst case, below operation costs O(N)
                for (int[][] ps : map.get(pm).get(dist2)) {
                    double d1 = dist2(p1, ps[0]);
                    double d2 = dist2(p1, ps[1]);
                    res = Math.min(res, Math.sqrt(d1 * d2));
                }
                map.get(pm).get(dist2).add(new int[][]{p1, p2});
            }
        }

        return res == Double.MAX_VALUE ? 0 : res;
    }

    private double dist2(int[] p1, int[] p2) {
        return (p2[0]-p1[0])*(p2[0]-p1[0]) + (p2[1]-p1[1])*(p2[1]-p1[1]);
    }
}